### PR TITLE
Implement Join Pruning with ORCA

### PR DESCRIPTION
**Pull request by DevChattopadhyay**
Current Implementation:
========================

```
create table foo (a int,b int,c int, d int,e int,constraint idx1 unique(a,b),constraint idx2 unique(a,c,d));
create table bar (p int,q int,r int, s int,t int,constraint idx3 unique(p,q),constraint idx4 unique(p,r,s));

explain select foo.* from foo left join bar on foo.a=bar.p and foo.c=bar.r and foo.e=bar.s where foo.b>100;

                                  QUERY PLAN
------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..12.00 rows=2 width=20)
   ->  Nested Loop Left Join  (cost=0.00..12.00 rows=1 width=20)
         Join Filter: true
         ->  Index Scan using idx1 on foo  (cost=0.00..6.00 rows=1 width=20)
               Index Cond: (b > 100)
         ->  Index Scan using idx3 on bar  (cost=0.00..6.00 rows=1 width=1)
               Index Cond: (p = foo.a)
               Filter: ((foo.c = r) AND (foo.e = s))
 Optimizer: Pivotal Optimizer (GPORCA)
(9 rows)

set optimizer to off;
explain select foo.* from foo left join bar on foo.a=bar.p and foo.c=bar.r and foo.e=bar.s where foo.b>100;
                                    QUERY PLAN
-----------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..597.17 rows=21800 width=20)
   ->  Seq Scan on foo  (cost=0.00..306.50 rows=7267 width=20)
         Filter: (b > 100)
 Optimizer: Postgres query optimizer
(4 rows)

```

After Fix:
==========

```
create table foo (a int,b int,c int, d int,e int,constraint idx1 unique(a,b),constraint idx2 unique(a,c,d));
create table bar (p int,q int,r int, s int,t int,constraint idx3 unique(p,q),constraint idx4 unique(p,r,s));

explain select foo.* from foo left join bar on foo.a=bar.p and foo.c=bar.r and foo.e=bar.s where foo.b>100;
                                QUERY PLAN
-----------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..6.00 rows=1 width=20)
   ->  Index Scan using idx1 on foo  (cost=0.00..6.00 rows=1 width=20)
         Index Cond: (b > 100)
 Optimizer: Pivotal Optimizer (GPORCA)
(4 rows)

```
Conditions for join pruning:
============================

1) The output columns and the WHERE clause should only contain columns from the outer relation. 
2) All the unique/primary keys from a key set collection of the inner relation should be present in the join condition. 
3) The above mentioned keys should either be constant or equal to a column from the outer relation.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 1]
```diff
@@ -3040,6 +3042,472 @@ CExpressionPreprocessor::ConvertSplitUpdateToInPlaceUpdate(CMemoryPool *mp,
 	return pexpr;
 }
 
+CExpression *
+CExpressionPreprocessor::PexprJoinPruningScalar(CMemoryPool *mp,
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3046]: Can you add a comment to explain what does each function do?

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 2]
```diff
@@ -3040,6 +3042,472 @@ CExpressionPreprocessor::ConvertSplitUpdateToInPlaceUpdate(CMemoryPool *mp,
...
+					present_in_inner_rel->Release();
+					result->Release();
+					inner_rel_unique_keys->Release();
+					return pexprNew;
+				}
+			}
+		}
+		else if (COperator::EopScalarBoolOp == join_cond->Pop()->Eopid() &&
+				 CScalarBoolOp::EboolopAnd ==
+					 CScalarBoolOp::PopConvert(join_cond->Pop())->Eboolop())
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3353]: CPredicateUtils::FAnd

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 3]
```diff
@@ -3040,6 +3042,472 @@ CExpressionPreprocessor::ConvertSplitUpdateToInPlaceUpdate(CMemoryPool *mp,
...
+		}
+
+		else
+		{
+			output_columns->Release();
+			childs_output_columns->Release();
+			result->Release();
+			inner_rel_unique_keys->Release();
+			return pexprNew;
+		}
```
Comment (gpopt) [inline src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3456]: This part should be optimized, duplicate.

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 4]
```diff
@@ -3040,6 +3042,551 @@ CExpressionPreprocessor::ConvertSplitUpdateToInPlaceUpdate(CMemoryPool *mp,
...
+				subquery_colref = subquery_pop->Pcr();
+			}
+			if (COperator::EopScalarSubqueryAll == pexprScalar->Pop()->Eopid())
+			{
+				CScalarSubqueryAll *subquery_pop =
+					CScalarSubqueryAll::PopConvert(pexprScalar->Pop());
+
+
+				subquery_colref = subquery_pop->Pcr();
+			}
```
Comment (chrishajas) [inline src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3140]: Can this code be condensed? Can we cast it to CScalar instead to retrieve Pcr()?

**Review by chrishajas — COMMENTED** [review]
I'd still like to see this refactored more if possible. It's quite a large function and is quite nested. Are there more parts we can break out into functions?

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 5]
```diff
@@ -3040,6 +3042,472 @@ CExpressionPreprocessor::ConvertSplitUpdateToInPlaceUpdate(CMemoryPool *mp,
...
+	else
+	{
+		pdrgpexpr->Release();
+		pexprScalar->AddRef();
+		return pexprScalar;
+	}
+}
+
+CExpression *
+CExpressionPreprocessor::PexprJoinPruning(CMemoryPool *mp, CExpression *pexpr,
```
[the remark this answers is not in the document]
Comment (DevChattopadhyay) [reply src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3277]: Added Description

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp (hunk 1 above)
Comment (DevChattopadhyay) [reply src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3138]: Comment Added

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp (hunk 2 above)
Comment (DevChattopadhyay) [reply src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp]: Done

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp (hunk 3 above)
Comment (DevChattopadhyay) [reply src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp]: Removed the duplicate code

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp (hunk 4 above)
Comment (DevChattopadhyay) [reply src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp]: CScalar doesn't have any method to retrieve the Pcr(). I have tried to condense CScalarSubqueryAny and CScalarSubqueryAll by casting it to there parent class CScalarSubqueryQuantified.

**Comment by 24nishant** [comment]
Great work Dev!!

I read all the 3 conditions, which help to implement the Join Prunning.
Writing my understanding of the same, please correct me if i am wrong .

Scenario : 

```
create table tab1 (a int,b int, primary key (a));
create table tab2 (a int,b int,c int, primary key (a,b));
insert into tab1 values (1,10), (2,20);
insert into tab2 values (1,10,100), (2,20,200), (3,30,300),(1,40,400);
```

nishants4=# select * from tab1;
 a | b
---+----
 1 | 10
 2 | 20
(2 rows)

nishants4=# select * from tab2;
 a | b  |  c
---+----+-----
 3 | 30 | 300
 1 | 10 | 100
 2 | 20 | 200
 1 | 40 | 400
(4 rows)

PRUNNING CONDITIONS

  1. The output columns and the WHERE clause should only contain columns from the outer relation.

Condition on output column is required, because if we have any column from inner table then, we need to have a scan on it also.
Condition on where clause is required because, if we have a condition on the on the inner table, then it might happen that no tuples are returned by the join
eg

No join Prunning

nishants4=# select tab1.a from tab1 left join tab2 on tab1.a=tab2.a and tab1.b=tab2.b where tab2.b>100;
 a
-\--
(0 rows)

nishants4=# select tab1.a from tab1 left join tab2 on tab1.a=tab2.a and tab1.b=tab2.b where tab2.b=100;
 a
-\--
(0 rows)

Join Prunning

nishants4=# select tab1.a from tab1 left join tab2 on tab1.a=tab2.a and tab2.b=100;
 a
-\--
 2
 1
(2 rows)

  2. All the unique/primary keys from a particular key set collection of the inner relation should be present in the join condition.

This condition helps to identify 1:1 relation between outer and inner table, and also to eliminate any duplicate tuples in the inner table

No join prunning

nishants4=# select tab1.a from tab1 left join tab2 on tab1.a=tab2.a;
 a
-\--
 1
 1
 2
(3 rows)

Join Prunning

nishants4=# select tab1.a from tab1 left join tab2 on tab1.a=tab2.a and tab1.b=tab2.b;
 a
-\--
 2
 1
(2 rows)


  3. The unique keys of the inner relation should either be constant or equal to a column from the outer relation.

This condition inturn helps to remove any duplicate values in the column, consider case where tab2.b>1 will result multiples
tuples, thus join pruning wont give correct result

join Prunning
nishants4=# explain select tab1.a from tab1 left join tab2 on tab1.a=tab2.a and tab2.b=1; 

No join prunning

as we have multiple tuple from tab2 which satisfies this condition.
nishants4=# explain select tab1.a from tab1 left join tab2 on tab1.a=tab2.a and tab2.b>1;

File: src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 6]
```diff
@@ -3040,6 +3042,479 @@ CExpressionPreprocessor::ConvertSplitUpdateToInPlaceUpdate(CMemoryPool *mp,
...
+	}
+}
+
+/*
+ This method is used to prune the CLogicalLeftOuterJoin operator in a bottom up approach
+ based on the following conditions
+ 1. The output columns and the WHERE clause should only contain columns from the outer relation.
+ 2. All the unique/primary keys from a particular key set collection of the inner relation should be present in the join condition.
+ 3. The unique keys of the inner relation should either be constant or equal to a column from the outer relation.
+ example:
```
Comment (24nishant) [inline src/backend/gporca/libgpopt/src/operators/CExpressionPreprocessor.cpp:3237]: Great work Dev!!

I analysed, all the 3 conditions, which help to implement the Join Pruning.
Writing my understanding of the same, please correct me if i am wrong .

Scenario :

```
create table tab1 (a int,b int, primary key (a));
create table tab2 (a int,b int,c int, primary key (a,b));
insert into tab1 values (1,10), (2,20);
insert into tab2 values (1,10,100), (2,20,200), (3,30,300),(1,40,400);
```
Please check the attached document for details.
[JoinPruning Conditions.pdf](https://github.com/greenplum-db/gpdb/files/10334601/JoinPruning.Conditions.pdf)

**Comment by DevChattopadhyay** [comment]
Closing this PR. Will raise a new PR in its place

**Pull request closed without merging**